### PR TITLE
[xDS] implement composite filter

### PR DESCRIPTION
**Pull request by markdroth**
Implements gRFC A103 (https://github.com/grpc/proposal/pull/511).

Note: This does not yet include the CEL matching support, which will be implemented separately.

The filter is implemented as a v3 interceptor.  In order to run this in the v1 stack, I have introduced a new adaptor called `V3InterceptorToV2Bridge` that allows running a v3 interceptor as a v2 filter, which will then sit on top of the PromiseBasedFilter code to run on the v1 stack.

Currently, we do not have a unit test framework for v3 interceptors, so the only tests here are end2end tests.  Once we have such a framework, we can write unit tests as well.

### CODE REVIEW

**Comment by markdroth** [comment]
I think this is now ready for review.

@ac-patel, please review the promise code to make sure it looks sane.  @ctiller, you may be interested in this as well, but feel free to skip the review if you don't have time.

@rishesh007, please review the composite filter logic.

Thanks!

File: src/core/xds/grpc/xds_http_composite_filter.cc [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+  switch (denominator) {
+    case envoy_type_v3_FractionalPercent_MILLION:
+      return numerator;
+    case envoy_type_v3_FractionalPercent_TEN_THOUSAND:
+      return numerator * 100;
+    case envoy_type_v3_FractionalPercent_HUNDRED:
+    default:
+      return numerator * 10000;
+  }
+}
```
Comment (rishesh007) [inline src/core/xds/grpc/xds_http_composite_filter.cc]: can we move this to some common util files, cause I need something similar for ext_authz and might be needed for some other feature

File: src/core/call/call_spine.h [hunk 2]
```diff
@@ -81,10 +81,14 @@ class CallSpine final : public Party, public channelz::DataSource {
   }
 
   auto PullServerInitialMetadata() {
+    GRPC_TRACE_LOG(call_state, INFO)
```
Comment (ac-patel) [inline src/core/call/call_spine.h:87]: Is there a specific reason to use `GRPC_TRACE_LOG` over `GRPC_TRACE_DLOG`?

File: src/core/lib/channel/promise_based_filter.h [hunk 3]
```diff
@@ -1220,6 +1223,223 @@ class ImplementChannelFilter : public ChannelFilter,
...
+              // Intercept all pipes from v2 API.
+              call_args.client_to_server_messages->InterceptAndMap(
+                  [initiator, handler,
+                   pipe_owner](MessageHandle message) mutable {
+                    // Push the message onto the v3 initiator in its activity.
+                    initiator.SpawnPushMessage(std::move(message));
+                    // In the v3 handler's activity, pull the messages.
+                    // Use an inter-activity pipe to get them back to
+                    // the v2 activity.
+                    handler.SpawnGuarded(
```
Comment (ac-patel) [inline src/core/lib/channel/promise_based_filter.h]: I think it might be better to just spawn a Loop promise doing the same logic in the handler once. This would reduce per message work significantly. Also spawning a long lived promise would be fine as this `CallSpine` will not have many spawned promises (note there is an upper limit of 16 promises per `CallSpine`).

File: src/core/lib/channel/promise_based_filter.h [hunk 4]
```diff
@@ -1220,6 +1223,223 @@ class ImplementChannelFilter : public ChannelFilter,
...
+                  });
+              call_args.server_initial_metadata->InterceptAndMap(
+                  [initiator, handler,
+                   pipe_owner](ServerMetadataHandle metadata) mutable {
+                    // Push the metadata onto the v3 handler in its activity.
+                    handler.SpawnPushServerInitialMetadata(std::move(metadata));
+                    // In the v3 initiator's activity, pull the metadata.
+                    // Use an inter-activity latch to get it back to
+                    // the v2 activity.
+                    initiator.SpawnGuarded(
```
Comment (ac-patel) [inline src/core/lib/channel/promise_based_filter.h]: Optional nit: This promise can also be spawned before-hand.

File: src/core/lib/channel/promise_based_filter.h [hunk 5]
```diff
@@ -1220,6 +1223,223 @@ class ImplementChannelFilter : public ChannelFilter,
...
+                  });
+              call_args.server_to_client_messages->InterceptAndMap(
+                  [initiator, handler,
+                   pipe_owner](MessageHandle message) mutable {
+                    // Push the message onto the v3 handler in its activity.
+                    handler.SpawnPushMessage(std::move(message));
+                    // In the v3 initiator's activity, pull the messages.
+                    // Use an inter-activity pipe to get them back to
+                    // the v2 activity.
+                    initiator.SpawnGuarded(
```
Comment (ac-patel) [inline src/core/lib/channel/promise_based_filter.h]: Same comment as the handler above.

File: src/core/xds/grpc/xds_common_types_parser.h [hunk 6]
```diff
@@ -58,7 +58,6 @@ inline std::optional<uint32_t> ParseUInt32Value(
   return google_protobuf_UInt32Value_value(proto);
 }
 
-// Returns the IP address in URI form.
```
Comment (rishesh007) [inline src/core/xds/grpc/xds_common_types_parser.h:62]: any reason why are we removing this?

File: src/core/call/call_spine.h (hunk 2 above)
Comment (markdroth) [reply src/core/call/call_spine.h:87]: The point of most tracers is to be able to turn them on in a production binary to get more logs when a problem occurs.  People don't tend to run debug binaries in production, so making this a debug-only trace log wouldn't really make it useful.

File: src/core/lib/channel/promise_based_filter.h (hunk 3 above)
Comment (markdroth) [reply src/core/lib/channel/promise_based_filter.h]: Good catch!  This promise is already handling all messages in the stream, so we don't need to spawn it more than once.  I've moved it outside of the `InterceptAndMap()`.

File: src/core/lib/channel/promise_based_filter.h (hunk 4 above)
Comment (markdroth) [reply src/core/lib/channel/promise_based_filter.h]: Done for consistency.

File: src/core/lib/channel/promise_based_filter.h (hunk 5 above)
Comment (markdroth) [reply src/core/lib/channel/promise_based_filter.h]: Done.

File: src/core/xds/grpc/xds_common_types_parser.h (hunk 6 above)
Comment (markdroth) [reply src/core/xds/grpc/xds_common_types_parser.h:62]: This is just something I happened to notice while working on this PR.  The comment is out of date: the function returns the address in sockaddr form, not URI fom.

File: src/core/xds/grpc/xds_http_composite_filter.cc (hunk 1 above)
Comment (markdroth) [reply src/core/xds/grpc/xds_http_composite_filter.cc]: Sure, done.  I also added some tests and fixed an edge case.

**Review by ac-patel — APPROVED** [review]
LGTM for the promise code
